From bab4935a6700ae722b1a05c193fad8254683fa06 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Wed, 23 Mar 2022 18:51:35 +0100
Subject: [PATCH 153/841] feat(st-ddr): add firmware completion wait procedure

Implement the user custom sequence that reads decodes firmware major
messages and follow the process until completion.
This part is extracted from the Boot Tool generation experience.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git b5fe1aeb6d2eb5e339af150d358f658c0177db70 feat(st-ddr): add firmware completion wait procedure]
Change-Id: I52f84e329aae63aa54497f15dd9d91810c5d468a
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/245745
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 .../ddrphy_phyinit_usercustom_g_waitfwdone.c  | 97 +++++++++++++++++++
 1 file changed, 97 insertions(+)

diff --git a/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c b/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c
index 1b1d0daec..b75bc7b23 100644
--- a/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c
+++ b/drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c
@@ -5,8 +5,80 @@
  */
 
 #include <common/debug.h>
+#include <drivers/delay_timer.h>
+#include <lib/mmio.h>
 
 #include <ddrphy_phyinit.h>
+#include <platform_def.h>
+
+/* Firmware major messages */
+#define FW_MAJ_MSG_TRAINING_SUCCESS	0x0000007
+#define FW_MAJ_MSG_START_STREAMING	0x0000008
+#define FW_MAJ_MSG_TRAINING_FAILED	0x00000FF
+
+#define PHYINIT_DELAY_1US		1U
+#define PHYINIT_DELAY_10US		10U
+
+static void wait_uctwriteprotshadow(bool state)
+{
+	uint16_t read_data;
+	uint16_t value = state ? BIT(0) : 0U;
+
+	do {
+		read_data = mmio_read_16((uintptr_t)(DDRPHYC_BASE +
+						     (4 * (TAPBONLY | CSR_UCTSHADOWREGS_ADDR))));
+		udelay(PHYINIT_DELAY_1US);
+	} while ((read_data & BIT(0)) != value);
+}
+
+static void ack_message_receipt(void)
+{
+	/* Acknowledge the receipt of the message */
+	mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4 * (TAPBONLY | CSR_DCTWRITEPROT_ADDR))), 0U);
+
+	udelay(PHYINIT_DELAY_1US);
+
+	wait_uctwriteprotshadow(true);
+
+	/* Complete the 4-phase protocol */
+	mmio_write_16((uintptr_t)(DDRPHYC_BASE + (4 * (TAPBONLY | CSR_DCTWRITEPROT_ADDR))), 1U);
+
+	udelay(PHYINIT_DELAY_1US);
+}
+
+static int get_major_message(void)
+{
+	int message_number;
+
+	wait_uctwriteprotshadow(false);
+
+	message_number = mmio_read_16((uintptr_t)(DDRPHYC_BASE +
+						  (4 * (TAPBONLY | CSR_UCTWRITEONLYSHADOW_ADDR))));
+
+	ack_message_receipt();
+
+	return message_number;
+}
+
+static int get_streaming_message(void)
+{
+	int stream_word_lower_part;
+	int stream_word_upper_part;
+
+	wait_uctwriteprotshadow(false);
+
+	stream_word_lower_part = mmio_read_16((uintptr_t)(DDRPHYC_BASE +
+							  (4 * (TAPBONLY |
+								CSR_UCTWRITEONLYSHADOW_ADDR))));
+
+	stream_word_upper_part = mmio_read_16((uintptr_t)(DDRPHYC_BASE +
+							  (4 * (TAPBONLY |
+								CSR_UCTDATWRITEONLYSHADOW_ADDR))));
+
+	ack_message_receipt();
+
+	return stream_word_lower_part | (stream_word_upper_part << 16);
+}
 
 /*
  * Implements the mechanism to wait for completion of training firmware execution.
@@ -29,7 +101,32 @@
  */
 void ddrphy_phyinit_usercustom_g_waitfwdone(void)
 {
+	int fw_major_message;
+
 	VERBOSE("%s Start\n", __func__);
 
+	do {
+		fw_major_message = get_major_message();
+		VERBOSE("fw_major_message = %x\n", (unsigned int)fw_major_message);
+
+		if (fw_major_message == FW_MAJ_MSG_START_STREAMING) {
+			int i;
+			int read_data = get_streaming_message();
+			int stream_len = read_data & 0xFFFF;
+
+			for (i = 0; i < stream_len; i++) {
+				read_data = get_streaming_message();
+				VERBOSE("streaming message = %x\n", (unsigned int)read_data);
+			}
+		}
+	} while ((fw_major_message != FW_MAJ_MSG_TRAINING_SUCCESS) &&
+		 (fw_major_message != FW_MAJ_MSG_TRAINING_FAILED));
+
+	udelay(PHYINIT_DELAY_10US);
+
+	if (fw_major_message == FW_MAJ_MSG_TRAINING_FAILED) {
+		ERROR("%s Training has failed.\n", __func__);
+	}
+
 	VERBOSE("%s End\n", __func__);
 }
-- 
2.47.2

